fixes #14666 - support IPv6 in UI

timogoebel opened the pull request.

timogoebel committed:
fixes #14666 - support IPv6 in UI

dLobatog commented on app/views/nic/_base_form.html.erb:
```diff
@@ -28,25 +30,47 @@
                 ).html_safe,
                 :size => "col-md-8", :label_size => "col-md-3" %>
 
+<%= text_f f, :ip6,
```
When I try to click on 'suggest new ip' , it returns 404 because I have not set a MAC yet. The error here should be a bit more informative than just 'not found'

timogoebel replied:
Note to self: When the mac is updated, a new ip is generated. We should maybe validate the mac and show errors early.

timogoebel replied:
@dLobatog : Implemented this. It still lacks a way what to do when the mac is generated by the compute resource.

timogoebel commented:
Rebased.

dLobatog commented on app/views/subnets/_fields.html.erb:
```diff
@@ -1,20 +1,25 @@
+  :disabled => !f.object.new_record? %>
 <%= text_f f, :network %>
-<%= text_f f, :mask, :help_inline => _("Netmask for this subnet") %>
+<%= text_f f, :cidr, :label => _("CIDR"), :help_inline => _("CIDR for this subnet") %>
```
This is the "prefix length", not the CIDR, the CIDR is `fc00:0:0:55::/64`, but this field refers to the 64 part, thus prefix length.

timogoebel replied:
@dLobatog: To be honest I don't want to argue about this. But If I'm not mistaken "prefix length" is an IPv6 term, for IPv4 it's called "suffix". CIDR just says, that you don't specify the subnet mask in dotted decimal notation, but you count the number of 1-bits in the subnet mask. That's why I chose CIDR, but I don't mind changing this.
I guess it would be better to show the subnet mask in dotted decimal notation for IPv4 and the prefix length for IPv6. That would require us to filter the opposite param in the controller. Do you think, that's worth the effort? Is there a better approach?

dLobatog commented on app/views/subnets/_fields.html.erb:
```diff
@@ -1,20 +1,25 @@
 <%= text_f f, :name %>
+<%= selectable_f f, :type,
+  options_for_select(Subnet.types_with_form_data, @subnet.type),
+  {:label => _("Protocol")},
```
I think you mean:

``` ruby
{ }, 
:label => _("Protocol"), :disabled...
```

Otherwise the label is ignored and it shows Type

dLobatog commented on app/views/nic/_base_form.html.erb:
```diff
@@ -28,25 +30,47 @@
+              :label => _("IPv6 address"),
+              :class => :interface_ip6,
+              :autocomplete => 'off',
+              :help_block => link_to(_("Suggest new"), '#', :class => 'suggest_new_ip6'),
```
This `Suggest new` should only be shown when that's actually possible: which is when `subnet.ipam` is `Internal DB`. Otherwise one can click w/o any feedback and it's confusing

dLobatog commented on app/services/ipam/dhcp.rb:
```diff
@@ -8,6 +8,10 @@ def suggest_ip
       logger.debug("Found #{ip}")
       ip
+    rescue => e
+      logger.warn "Failed to fetch a free IP from our proxy: #{e}"
```
Can you also add the proxy name and URL in here? It's a bit of a pain to have to click on subnet and check what proxy is associated with the subnet when you're creating a host

tbrisker commented on app/helpers/hosts_and_hostgroups_helper.rb:
```diff
@@ -48,8 +50,8 @@ def accessible_puppet_proxies(obj)
-  def domain_subnets
-    accessible_related_resource(@domain, :subnets)
+  def domain_subnets(type)
+    accessible_related_resource(@domain, :subnets, :query => {:type => type})
```
Why not `accessible_related_resource(@domain, :subnets).where(:type=> type)` and then there is no need for changing the `accessible_related_resource` function?

timogoebel replied:
@tbrisker: I couldn't do it because `return [] if obj.blank?` would lead to `[].where(:type => type)` and that obviously does not work. Is there a better approach?

tbrisker replied:
aha, I see. that makes sense, could you please only rename the option from `:query` to `:where` so it is clear this is passed as an additional `where` to the filter? Also, please extract the options from the hash only after the `blank?` short circuit, no point in extracting them if they won't be used.

timogoebel replied:
@tbrisker : Sure, done. 👍

timogoebel commented:
@dLobatog : Addressed your comments. Added hiding of the "suggest new links" when the subnet does not support suggesting a new ip.

dLobatog commented:
@timogoebel I'm a bit confused about the CIDR vs prefix length discussion in https://github.com/theforeman/foreman/pull/3587#discussion_r69083574 . What did you want that field to contain? You can also use a subnet mask in dotted hexadecimal notation for IPv6 but I don't think it's common vs using a suffix. 

`That would require us to filter the opposite param in the controller. Do you think, that's worth the effort? Is there a better approach?`

This bit confused me too, what's 'the opposite param in the controller'?

timogoebel commented:
@dLobatog 

> I'm a bit confused about the CIDR vs prefix length discussion in #3587 (comment) .

Sorry for not making this clear. I meant to say the following:
- Currently we use the subnet mask for IPv4 (255.255.255.0), which is fine.
- For IPv6 the subnet mask is not used as you said (ffff:ffff:ffff:ffff:0000:0000:0000:0000)
- The current status of this patch changes both IPv4 and IPv6 to use the "cidr"/prefix ("24" as in 192.168.1.1/24; "64" as in 2001:db8::1/64)
- We could change the UI to use "255.255.255.255" for IPv4 and "64" for IPv6.

My first thought on how to implement the latter is to just toggle visibility of the prefix and the mask field via jQuery. The browser than still sends both information in the request. That's why I meant we need to filter it in the controller. There might be a better approach, haven't come up with something though.

dLobatog commented on app/services/ipam/base.rb:
```diff
@@ -1,23 +1,40 @@
+      errors.add(:mac, _('is not a valid MAC address'))
+    end
+
+    def read_attribute_for_validation(attr)
```
I had to look up what this does, mind to add a comment explaining why this is ncessary?

dLobatog commented on app/services/ipam/base.rb:
```diff
@@ -1,23 +1,40 @@
+      true
+    end
+
+    def serializable_hash(options = nil)
```
When is this serialized?

timogoebel replied:
https://github.com/timogoebel/foreman/blob/cdb33fed2b996e6e5fa4d6b610e5ff3a7a449260/app/controllers/concerns/foreman/controller/host_details.rb#L27-L29

`#domain_selected`  in `app/controllers/concerns/foreman/controller/host_details.rb` calls `as_json`, this needs the `serializable_hash` method.

dLobatog commented on app/services/ipam/base.rb:
```diff
@@ -1,23 +1,40 @@
 module IPAM
   class Base
+    extend ActiveModel::Naming
```
Sorry, what are these two for?

timogoebel replied:
They are required for Active Model Errors, see http://api.rubyonrails.org/classes/ActiveModel/Errors.html

dLobatog commented on app/assets/javascripts/host_edit.js:
```diff
@@ -550,40 +544,70 @@ $(document).on('change', '#host_provision_method_image', image_provision_method_
+  e.preventDefault();
+});
+
+$(document).on('click', '.suggest_new_ip6', function (e) {
```
Duplicate code, it's same as ipv4 except for 1 parameter

dLobatog commented on app/assets/javascripts/host_edit.js:
```diff
@@ -550,40 +544,70 @@ $(document).on('change', '#host_provision_method_image', image_provision_method_
+  interface_subnet_selected(this, 'ip');
+});
+
+$(document).on('change', '.interface_subnet6', function () {
```
Duplicate code

timogoebel replied:
Any good idea on how to refactor this out?

dLobatog commented on app/assets/javascripts/subnets.js:
```diff
@@ -1,10 +1,43 @@
+    $("#subnet_dhcp_id option[value='']").attr("selected","selected");
+  }
+
+  if (element.find(':selected').data('supports_tftp')) {
```
duplicate code (dhcp/tftp)

dLobatog commented on app/models/subnet/ipv6.rb:
```diff
@@ -26,6 +26,10 @@ def self.supported_ipam_modes
   end
 
+  def self.supports_tftp?
+    false
```
Why is tftp not supported? xinetd / tftpd which come with the foreman installer support IPv6

timogoebel replied:
They do, but Foreman currently does not support this. IMHO tftp orchestration only makes sense in combination with dhcp orchestration. We need efi support in Foreman first before dhcp orchestration can work. So this is still a lot more work to do.

dLobatog replied:
Sorry, I still don't understand, TFTP orchestration is independent of DHCP, in fact some people do static network provisioning with Foreman (http://serverfault.com/questions/609770/using-foreman-to-provision-a-vm-with-static-ip for instance). The EFI part caught me completely off guard, why is that necessary?

timogoebel replied:
@dLobatog : Just a short answer for now, don't have a lot of time today:

For tftp orchestration to work, we need to make changes to `app/models/concerns/orchestration/tftp.rb` before it can work for ipv6. There should be "subnet6" somewhere, haven't given it much thought right now. My main problem is that it's actually IPv4 and IPv6 independent. So when a host has an IPv4 subnet and a IPv6 subnet and both have a tftp proxy set, which one do you choose. That's why I wanted to hide it in the UI for now. We can easily add it later again when it's properly tested.

To provision a server with ipv6 only you would need the bios to make a dhcpv6 request. Most bioses don't support this. Most EFI firmwares do though. That's why I wrote we need EFI support. For VMWare you have to tell the compute resource to do an EFI boot and explicity set it to boot via IPv6. fog-vsphere doesn't support this right now. So we'd (read: I'd) have to add this first.

Does that make sense to you? If you want, we can do a hangout and discuss things there.

dLobatog commented on app/services/ipam/eui64.rb:
```diff
@@ -1,16 +1,27 @@
+    def validate
+      errors.add(:mac, _("can't be blank")) unless mac.present?
+      errors.add(:subnet, _("network can't be blank")) unless subnet.network.present?
+      errors.add(:subnet, _("CIDR can't be blank")) unless subnet.cidr.present?
```
"did you mean prefix length?" - ![clippy](https://images.duckduckgo.com/iu/?u=http%3A%2F%2Fcdn.geekwire.com%2Fwp-content%2Fuploads%2F2011%2F04%2Fclippy.jpg&f=1)

dLobatog commented on .rubocop_todo.yml:
```diff
@@ -89,7 +89,7 @@ Metrics/MethodLength:
 # Configuration parameters: CountComments.
 Metrics/ModuleLength:
-  Max: 408
+  Max: 431
```
Can you find some stuff to split out in the helper so that it doesn't go beyond the limit? https://github.com/theforeman/foreman/pull/3621/files I had to do that here , if that works for you feel free to take it

dLobatog commented:
@timogoebel 

> ```
> The current status of this patch changes both IPv4 and IPv6 to use the "cidr"/prefix ("24" as in 192.168.1.1/24; "64" as in 2001:db8::1/64) 
> ```

That's only because the field that used to contain 'mask' now it's set to 'cidr'. What you said before about toggling visibility of the prefix and mask makes sense IMO. Especially since the mask field is used **everywhere** in Foreman. 

The way I envision it is:

ipv4 - displays mask, cidr field is ignored (or set from the mask?, just trying to make it easy for now)
ipv6 - displays prefix, mask field is ignored 

Makes sense?

timogoebel commented:
> The way I envision it is:
> 
> ipv4 - displays mask, cidr field is ignored (or set from the mask?, just trying to make it easy for now)
> ipv6 - displays prefix, mask field is ignored

@dLobatog : Addressed most of your comments. Your vision became reality. :-P Had to add some client side validation for this to work. Mind to test the subnet form? Happy to hear your thoughts.

dLobatog commented on bundler.d/assets.rb:
```diff
@@ -16,3 +16,4 @@
 gem 'spice-html5-rails', '~> 0.1.5'
 gem 'quiet_assets', '~> 1.0'
 gem 'jquery_pwstrength_bootstrap_4', :require => 'jquery_pwstrength_bootstrap'
+gem 'ipaddrjs-rails', '~> 1.1.1'
```
This needs to be packaged, @timogoebel let me know if you want me to do it or you prefer to do it. The README in https://github.com/theforeman/foreman-packaging/tree/rpm/develop explains how to do this, it's very simple (run gem2rpm , update comps/ and rel-eng/tito.props and submit the PR :smile: )

timogoebel replied:
No worries, I'll do it tomorrow. :-)

dLobatog commented:
@timogoebel Thanks, the latest updates look great. I can't find anything I don't agree with, or that doesn't work on the UI, so I'm merging it. Please respond to https://github.com/theforeman/foreman/pull/3587#discussion_r70662494 :smile: 

Thanks again @timogoebel , it's been a lot of work!

dLobatog commented:
Merged as 79c0664a008afe0fea54a0b5256426f3867a02bc, thanks @timogoebel!
